File: backend/src/utils/crypto.ts
import crypto from 'crypto';
import logger from './logger';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 16;
const TAG_LENGTH = 16;

function getKey(): Buffer {
  const keyHex = process.env.ENCRYPTION_KEY;
  if (!keyHex) throw new Error('ENCRYPTION_KEY environment variable is not set');
  const key = Buffer.from(keyHex, 'utf8');
  if (key.length !== 32) {
    throw new Error('ENCRYPTION_KEY must be exactly 32 bytes (UTF-8)');
  }
  return key;
}

export function encrypt(plaintext: string): string {
  logger.debug('Encrypting token');
  const key = getKey();
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);
  const encrypted = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  const result = Buffer.concat([iv, tag, encrypted]).toString('base64');
  logger.debug('Token encrypted successfully');
  return result;
}

export function decrypt(ciphertext: string): string {
  logger.debug('Decrypting token');
  const key = getKey();
  const data = Buffer.from(ciphertext, 'base64');
  const iv = data.subarray(0, IV_LENGTH);
  const tag = data.subarray(IV_LENGTH, IV_LENGTH + TAG_LENGTH);
  const encrypted = data.subarray(IV_LENGTH + TAG_LENGTH);
  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(tag);
  const result = Buffer.concat([decipher.update(encrypted), decipher.final()]).toString('utf8');
  logger.debug('Token decrypted successfully');
  return result;
}

File: importers/core/crypto.py
from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from core.logger import get_logger

logger = get_logger("crypto")

IV_LENGTH = 16
TAG_LENGTH = 16  # GCM auth tag is appended by AESGCM automatically


def _get_key() -> bytes:
    key_str = os.environ.get("ENCRYPTION_KEY")
    if not key_str:
        raise RuntimeError("ENCRYPTION_KEY environment variable is not set")
    key = key_str.encode("utf-8")
    if len(key) != 32:
        raise RuntimeError("ENCRYPTION_KEY must be exactly 32 bytes (UTF-8)")
    return key


def decrypt_token(ciphertext_b64: str) -> str:
    """Decrypt a token that was encrypted by the Node.js backend using AES-256-GCM."""
    logger.debug("Decrypting token")
    key = _get_key()
    data = base64.b64decode(ciphertext_b64)
    iv = data[:IV_LENGTH]
    # Node.js crypto appends the auth tag separately; layout is: iv(16) | tag(16) | ciphertext
    tag = data[IV_LENGTH : IV_LENGTH + TAG_LENGTH]
    ciphertext = data[IV_LENGTH + TAG_LENGTH :]
    # AESGCM expects ciphertext + tag appended
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
    logger.debug("Token decrypted successfully")
    return plaintext.decode("utf-8")
